takashi-task-framework.md

```diff
@@ -0,0 +1,265 @@
+
+```ini
+[RunIf]
+Requires="./check-if-too-many-users.sh" # checks for user count of all
+accounts, returns 0 if the task needs to run, 1 otherwise.
+
+[Command]
+command="./prune-users.sh"
+```
+
+Otherwise we can support a cron-like syntax:
+
+```ini
+[RunIf]
+cron="* * * * *" # run every minute
+
+[Command]
+command="./heartbeat.sh" # check if the other server is still up.
+```
+
+We should also support a way to say how many times to restart a service
+(if one is flaky, maybe we want to try 5 times). We also want to be able
+to provide memory and disk limits, because otherwise tasks might (if
+coded improperly) take down other tasks, which would be bad.
+
+```ini
+[Service]
+Restart=5 # restart 5 times at most.
+LimitMem=8G # allocate 8GB for this task.
+LimitDisk=4G # allocate 4GB of memory for this task.
+```
+
+We can also "jail" each task, by using a FreeBSD-esque jail. Each task
+will be granted its own process space, memory, and disk, and cannot
+escape (in case malicious or fault code is executed, it won't be able to
+cat /etc/passwd or other evil things).
+
+## Logging
+
+We need to make sure that logging is prioritized, since this is
+complicated. If I write a task file and one of the dependencies fails to
+install, this is a problem as well. We'll surface a command like
+`task logs [$NAME]`
+which will let the user query the task runner for the stdout of a task name.
+
+## Running at Scale
+
+Since we can assume that each task is idempotent (we'll jail each task
+and assume they can run in any order, including in parallel), we can
+allocate a cluster of machines to run tasks for us, and there can be a
+centralized metadata service that users can query for the status of the
+jobs they care about, along with audits of status changes, which they
+can query by running `task audit [NAME]`.
+
+We can assume that running 1B tasks per day is about 10k tasks per
+second. We can try to keep a cluster of a thousand machines to run jobs,
+all responding back to a centralized system that is replicated to store
+metadata on tasks. The metadata server can generate ids for each task,
+and also queries the tasks that need to run (cron jobs) and runs the
+dependencies of all tasks provided, to know when to run a task if its
+`RunIf` block wants it to run.
+
+The centralized system could use a backing store of a time-series like
+database, or it could use an append-only system, where it appends each
+event that happens to a table, and a query will be indexed to make sure
+it stays efficient (instead of table scanning).
+
+We can create a status table that contains enums like so:
+(this allows us to add more statuses in the future, just like errno in C).
+
+```sql
+create table status (
+  id primary key auto increment,
+  name varchar(255)
+);
+```
+
+```json
+{ "id": 1, "job_id": 3252, "status": 1, "timestamp": 17281 } # job 3252 completed at time 17281
+{ "id": 2, "job_id": 3253, "status": 2, "timestamp": 17283 } # job 3253
+is in progress
+```
```

Comment:
I see you use this pattern where a number represents a string in some enum so like status created is 1 status pending is 2. Do you ever see this becoming confusing for the end user who will have to keep track of all the status codes as our framework expands?

Reply:
Just above there's a definition for an SQL table: 

```sql
create table status (
  id primary key auto increment,
  name varchar(255)
);
```

In the actual data/source of truth, i'll default to using IDs, because they might change in the future, so I don't want to embed the textual representation (like created/pending/completed/failed) in that data.

For the end user I could do a lookup into the SQL table and use that to fill in the status when the data is queried.

So

```json
{ "id": 1, "job_id": 3252, "status": 1, "timestamp": 17281 }
```

Would become:

```json
{ "id": 1, "job_id": 3252, "status": "Completed", "timestamp": $UTC_TIMESTAMP }
```

Reply:
Just above there's a definition for an SQL table: 

```sql
create table status (
  id primary key auto increment,
  name varchar(255)
);
```

In the actual data/source of truth, i'll default to using IDs, because they might change in the future, so I don't want to embed the textual representation (like created/pending/completed/failed) in that data.

For the end user I could do a lookup into the SQL table and use that to fill in the status when the data is queried.

So

```json
{ "id": 1, "job_id": 3252, "status": 1, "timestamp": 17281 }
```

Would become:

```json
{ "id": 1, "job_id": 3252, "status": "Completed", "timestamp": $UTC_TIMESTAMP }
```